src/vsr/grid_repair_queue.zig

```diff
@@ -0,0 +1,489 @@
+        /// block-by-block.
+        content_blocks_received: TableContentBlocksSet = TableContentBlocksSet.initEmpty(),
+        /// This count includes the index block.
+        /// Invariants:
+        /// - table_blocks_written ≤ table_blocks_total
+        table_blocks_written: usize = 0,
+        /// When null, the table is awaiting an index block.
+        /// When non-null, the table is awaiting content blocks.
+        /// This count includes the index block.
+        table_blocks_total: ?usize = null,
+
+        /// Invoked when the table index block and all content blocks have been written.
+        callback: *const fn (*RepairTable, RepairTableResult) void,
+        /// "next" belongs to the `faulty_tables` FIFO.
+        next: ?*RepairTable = null,
+    };
+
+    pub const Options = struct {
+        /// Lower-bound for the limit of concurrent request_block()'s available.
+        blocks_max: usize,
+        /// Maximum number of concurrent request_table()'s.
+        tables_max: usize,
+    };
+
+    options: Options,
+
+    /// Invariants:
+    /// - For every block address in faulty_blocks, ¬free_set.is_free(address).
+    faulty_blocks: FaultyBlocks,
+    /// Index within `faulty_blocks`, used to cycle through block-repair requests.
+    ///
+    /// Invariants:
+    /// - faulty_blocks.count() > 0 implies faulty_blocks_repair_index < faulty_blocks.count()
+    /// - faulty_blocks.count() = 0 implies faulty_blocks_repair_index = faulty_blocks.count()
+    faulty_blocks_repair_index: usize = 0,
+
+    /// Invariants:
+    /// - For every index address in faulty_tables: ¬free_set.is_free(address).
+    faulty_tables: FIFO(RepairTable) = .{ .name = "grid_repair_queue_tables" },
+
+    checkpointing: ?struct {
+        /// The number of faulty_blocks with state=aborting.
+        aborting: usize,
+    } = null,
+
+    /// Guard against new block/table repairs queueing within RepairTable callbacks.
+    canceling: bool = false,
+
+    pub fn init(allocator: std.mem.Allocator, options: Options) error{OutOfMemory}!GridRepairQueue {
+        var faulty_blocks = FaultyBlocks{};
+        errdefer faulty_blocks.deinit(allocator);
+
+        try faulty_blocks.ensureTotalCapacity(
+            allocator,
+            options.blocks_max + options.tables_max * lsm_table_content_blocks_max,
+        );
+
+        return GridRepairQueue{
+            .options = options,
+            .faulty_blocks = faulty_blocks,
+        };
+    }
+
+    pub fn deinit(queue: *GridRepairQueue, allocator: std.mem.Allocator) void {
+        assert(!queue.canceling);
+
+        queue.faulty_blocks.deinit(allocator);
+
+        queue.* = undefined;
+    }
+
+    /// When the queue wants more blocks than fit in a single request message, successive calls
+    /// to this function cycle through the pending BlockRequests.
+    pub fn block_requests(queue: *GridRepairQueue, requests: []vsr.BlockRequest) usize {
+        assert(!queue.canceling);
+
```

Comment:
It might be empty if all of the requests were taking by the `read_global_queue`:

```zig
        pub fn block_requests(grid: *Grid, requests: []vsr.BlockRequest) usize {
            const request_faults_count = @min(grid.read_global_queue.count, requests.len);
            const request_repairs_count =
                grid.repair_queue.block_requests(requests[request_faults_count..]);
```

Reply:
Ah, then we can assert this in `Grid.block_requests`! I am actually unsure here --- on the one hand, empty slice is a natural neutral element here, on the other hand, we should never issue "empty" network messages or "empty" disk writes, and deep asserts seem useful.